packages/next/src/server/next.ts

```diff
@@ -42,26 +43,69 @@ export type NextServerOptions = Omit<
 > &
   Partial<Pick<ServerOptions | DevServerOptions, 'conf'>>
 
-export interface RequestHandler {
-  (
-    req: IncomingMessage,
-    res: ServerResponse,
-    parsedUrl?: NextUrlWithParsedQuery | undefined
-  ): Promise<void>
-}
+export type RequestHandler = (
+  req: IncomingMessage,
+  res: ServerResponse,
+  parsedUrl?: NextUrlWithParsedQuery | undefined
+) => Promise<void>
+
+export type UpgradeHandler = (
+  req: IncomingMessage,
+  socket: Duplex,
+  head: Buffer
+) => Promise<void>
 
 const SYMBOL_LOAD_CONFIG = Symbol('next.load_config')
 
-export class NextServer {
-  private serverPromise?: Promise<Server>
-  private server?: Server
+export interface NextWrapperServer {
+  // NOTE: the methods/properties here are the public API for custom servers.
+  // Consider backwards compatibilty when changing something here!
+
+  options: NextServerOptions
+  hostname: string | undefined
+  port: number | undefined
+
+  getRequestHandler(): RequestHandler
+  prepare(serverFields?: ServerFields): Promise<void>
+  setAssetPrefix(assetPrefix: string): void
+  close(): Promise<void>
+
+  // used internally
+  getUpgradeHandler(): UpgradeHandler
+
+  // legacy methods that we left exposed in the past
```

Comment:
see https://github.com/vercel/next.js/pull/73021 for some preparations to clean this up